rust/src/composepost.rs

```diff
@@ -549,10 +552,153 @@ pub fn composepost_nsswitch_altfiles(rootfs_dfd: i32) -> CxxResult<()> {
+) -> CxxResult<()> {
+    let rootfs = crate::ffiutil::ffi_view_openat_dir(rootfs_dfd);
+    let cancellable = &cancellable.gobj_wrap();
+
+    // TODO(lucab): unify this logic with the one in rpmostree-importer.cxx.
+    var_to_tmpfiles(&rootfs, Some(cancellable))?;
+    Ok(())
+}
+
+#[context("Converting /var to tmpfiles.d")]
+fn var_to_tmpfiles(rootfs: &openat::Dir, cancellable: Option<&gio::Cancellable>) -> Result<()> {
+    /* List of files that are known to possibly exist, but in practice
+     * things work fine if we simply ignore them.  Don't add something
+     * to this list unless you've verified it's handled correctly at
+     * runtime.  (And really both in CentOS and Fedora)
+     */
+    static KNOWN_STATE_FILES: &[&str] = &[
+        // https://bugzilla.redhat.com/show_bug.cgi?id=789407
+        "var/lib/systemd/random-seed",
+        "var/lib/systemd/catalog/database",
+        "var/lib/plymouth/boot-duration",
+        // These two are part of systemd's var.tmp
+        "var/log/wtmp",
+        "var/log/btmp",
+    ];
+
+    let pwdb = PasswdDB::populate_new(rootfs)?;
+
+    // We never want to traverse into /run when making tmpfiles since it's a tmpfs
+    // Note that in a Fedora root, /var/run is a symlink, though on el7, it can be a dir.
+    // See: https://github.com/projectatomic/rpm-ostree/pull/831
+    rootfs
+        .remove_all("var/run")
+        .context("Failed to remove /var/run")?;
+
+    // Here, delete some files ahead of time to avoid emitting warnings
+    // for things that are known to be harmless.
+    for path in KNOWN_STATE_FILES {
+        rootfs
+            .remove_file_optional(*path)
+            .with_context(|| format!("unlinkat({})", path))?;
+    }
+
+    // Convert /var wholesale to tmpfiles.d. Note that with unified core, this
+    // code should no longer be necessary as we convert packages on import.
+    // Make output file world-readable, no reason why not to
+    // https://bugzilla.redhat.com/show_bug.cgi?id=1631794
+    rootfs.ensure_dir_all("usr/lib/tmpfiles.d", 0o755)?;
+    rootfs.write_file_with_sync(
+        "usr/lib/tmpfiles.d/rpm-ostree-1-autovar.conf",
+        0o644,
+        |bufwr| -> Result<()> {
+            convert_path_to_tmpfiles_d_recurse(bufwr, &pwdb, &rootfs, "var", &cancellable)?;
+            Ok(())
+        },
+    )?;
+
+    Ok(())
+}
+
+/// Recursively explore target directory and translate content to tmpfiles.d entries.
+///
+/// This proceeds depth-first and progressively delete translated subpaths as it goes.
+#[context("Exploring /{} content", prefix)]
+fn convert_path_to_tmpfiles_d_recurse(
+    tmpfiles_bufwr: &mut BufWriter<File>,
+    pwdb: &PasswdDB,
+    rootfs: &openat::Dir,
+    prefix: &str,
+    cancellable: &Option<&gio::Cancellable>,
+) -> Result<()> {
+    use openat::SimpleType;
+
+    for subpath in rootfs.list_dir(prefix)? {
+        if cancellable.map(|c| c.is_cancelled()).unwrap_or_default() {
+            return Ok(());
```

Comment:
Ah thanks, that was actually I doubt I had. Is the cancel + error-bubbling race handle in some special way at the top level? Should this use some well-known marker error to signal that this didn't incur an actual failure?

Reply:
In practice, it doesn't matter here because we don't use the cancellable stuff on the compose side. 

On the client code the idea is the way things would work is that if you ctrl-c (SIGINT) the client then it sends a `Cancel` DBus method to the daemon, which flips the cancellable.  That would cause a cancellation error to propagate up the transaction thread, which would return it back to the client as a DBus error.

And today the client will just print `error: Cancelled` but it *could* detect that it was a "cancelled" error and treat that as "successfully cancelled".

In Rust with `Future<T>` cancelling is just dropping the task, which avoids this need at all.